lib/ecto/repo/queryable.ex

```diff
@@ -135,6 +135,15 @@ defmodule Ecto.Repo.Queryable do
     {{left, right}, values}
   end
 
+  defp transform_row({:%{}, _, pairs}, from, values) do
+    {result, values} = Enum.map_reduce pairs, values, fn({k, v}, acc) ->
```

Comment:
You can simply reduce here building both the map and retrieving the values in one pass!